### PR TITLE
Handle GC triggering in the constructor

### PR DESCRIPTION
**Pull request by casperisfine**
We've been seeing crashes caused by `stack_frames`:

```
    /app/components/platform/app/models/current_component.rb:5: [BUG] Segmentation fault at 0x0000177a00000000
    -- C level backtrace information -------------------------------------------
    /usr/local/ruby/bin/ruby(rb_print_backtrace+0x11) [0x562baad1bfae] vm_dump.c:759
    /usr/local/ruby/bin/ruby(rb_vm_bugreport) vm_dump.c:1045
    /usr/local/ruby/bin/ruby(rb_bug_for_fatal_signal+0xec) [0x562baadc90cc] error.c:821
    /usr/local/ruby/bin/ruby(sigsegv+0x4d) [0x562baac6fbad] signal.c:964
    /lib/x86_64-linux-gnu/libpthread.so.0(__restore_rt+0x0) [0x7f8e692b03c0]
    /usr/local/ruby/bin/ruby(RVALUE_MARKED+0x1c) [0x562baab5c347] gc.c:1509
    /usr/local/ruby/bin/ruby(gc_mark_set) gc.c:6581
    /usr/local/ruby/bin/ruby(gc_mark_ptr) gc.c:6695
    /tmp/bundle/ruby/3.1.0/gems/stack_frames-0.1.0/lib/stack_frames/stack_frames.so(buffer_mark+0x51) [0x7f8e5278e531] buffer.c:19
    /usr/local/ruby/bin/ruby(gc_mark_children+0x6e7) [0x562baab5df37] gc.c:6963
    /usr/local/ruby/bin/ruby(gc_mark_stacked_objects+0x3c) [0x562baab61b02] gc.c:7067
    /usr/local/ruby/bin/ruby(gc_mark_stacked_objects_incremental) gc.c:7101
    /usr/local/ruby/bin/ruby(gc_marks_step) gc.c:8064
    /usr/local/ruby/bin/ruby(gc_marks_continue) gc.c:8126
    /usr/local/ruby/bin/ruby(heap_prepare+0x4e) [0x562baab620fe] gc.c:2192
    /usr/local/ruby/bin/ruby(heap_next_freepage) gc.c:2384
    /usr/local/ruby/bin/ruby(ractor_cache_slots) gc.c:2420
    /usr/local/ruby/bin/ruby(newobj_slowpath) gc.c:2480
    /usr/local/ruby/bin/ruby(newobj_slowpath_wb_unprotected) gc.c:2506
    /usr/local/ruby/bin/ruby(newobj_of0+0x12) [0x562baab63aa2] gc.c:2542
    /usr/local/ruby/bin/ruby(newobj_of) gc.c:2551
    /usr/local/ruby/bin/ruby(rb_data_typed_object_wrap) gc.c:2761
    /usr/local/ruby/bin/ruby(rb_data_typed_object_zalloc) gc.c:2767
    /tmp/bundle/ruby/3.1.0/gems/stack_frames-0.1.0/lib/stack_frames/stack_frames.so(stack_frame_new+0x2a) [0x7f8e5278ebda] frame.c:34
    /tmp/bundle/ruby/3.1.0/gems/stack_frames-0.1.0/lib/stack_frames/stack_frames.so(buffer_initialize+0x9e) [0x7f8e5278e75e] buffer.c:65
```

What I think happens is that the GC triggers when `buffer->profile_frames` and `buffer->frames` is uninitialized.
So when we enter `buffer_mark`, we end up calling `rb_gc_mark` with uninitialized memory.

To fix this I use two things:

  - First use `ZALLOC_N` so that the memory is initialized to `NULL`, and check the addresses before calling `rb_gc_mark`
  - Second I initialize `buffer->capacity` and `buffer->lenght` before doing any further allocation.

One of the should suffice, but I guess it doesn't hurt to do both.

### CODE REVIEW

File: ext/stack_frames/buffer.c [hunk 1]
```diff
@@ -56,11 +61,14 @@ static VALUE buffer_initialize(VALUE self, VALUE size) {
     }
 
     TypedData_Get_Struct(self, buffer_t, &buffer_data_type, buffer);
-    buffer->profile_frames = ALLOC_N(VALUE, capacity);
-    buffer->lines = ALLOC_N(int, capacity);
-    buffer->frames = ALLOC_N(VALUE, capacity);
 
     buffer->capacity = 0;
+    buffer->length = 0;
```
Comment (peterzhu2118) [inline ext/stack_frames/buffer.c:66]: Doesn't [`TypedData_Make_Struct` allocate zero'd memory](https://github.com/ruby/ruby/blob/46f6575157d4c2f6bbd5693896e26a65037e5552/include/ruby/internal/core/rtypeddata.h#L458) in the data struct already? Would setting it to 0 here make a difference?

File: ext/stack_frames/buffer.c [hunk 2]
```diff
@@ -11,12 +11,17 @@ typedef struct {
...
 
     for (int i = 0; i < buffer->length; i++) {
-        rb_gc_mark(buffer->profile_frames[i]);
+        if (RTEST(frame = buffer->profile_frames[i])) {
+            rb_gc_mark(frame);
+        }
     }
     for (int i = 0; i < buffer->capacity; i++) {
-        rb_gc_mark(buffer->frames[i]);
+        if (RTEST(frame = buffer->frames[i])) {
```
Comment (peterzhu2118) [inline ext/stack_frames/buffer.c:22]: Looking at the code in `buffer_initialize`, shouldn't `buffer->frames[i]` never be null from `0 <= i < buffer->capacity`?

Comment (casperisfine) [reply ext/stack_frames/buffer.c:22]: Yep, as said in the PR description, it's redundant, but I wanted to be extra sure :p 

But yeah I could drop that extra check. I'm not even sure if it's a problem to `rb_gc_mark(Qnil)`.

File: ext/stack_frames/buffer.c (hunk 1 above)
Comment (casperisfine) [reply ext/stack_frames/buffer.c:66]: Ok, so both lines can go away? I'll simplify the PR.

**Comment by dylanahsmith** [comment]
Looks like this patch was already deployed, so I would be curious if it actually resolves the crashes you were seeing.

There had previously been a bug with memory unsafety from GC happening during buffer allocation and a regression test was added for that (https://github.com/Shopify/stack_frames/blob/7bd007acf5229df86ec2aefe229e74dc3ffe22ef/test/stack_frames/buffer_test.rb#L111-L116), so if there is still a bug with that (which I don't see for reasons @peterzhu2118 pointed out), then I'd be curious as to why that test isn't failing.  If it is relying on undefined behaviour, perhaps we should be testing with an address sanitizer to see if that will show the problem.

**Comment by casperisfine** [comment]
> There had previously been a bug with memory unsafety from GC

🤦 I forgot to look at unreleased changes. We were running the public `0.1.0`, seem like you already fixed the bug we're seeing, it just wasn't released. I'll try to confirm.

**Comment by casperisfine** [comment]
Yep, that's totally the same bug @XrXr already fixed in https://github.com/Shopify/stack_frames/pull/12.

Ok, I'll close and prepare a new release then.

**Pull request closed without merging**